### ISSUE TITLE
CUDA platform error: Error loading CUDA module: CUDA_ERROR_INVALID_PTX (218)

### ISSUE THREAD

**Issue by jmkoral** [openmm/openmm#3474]
I am very new to OpenMM and have struggled installing from source code, so I tried installing using Conda and have gotten much further, but am now receiving error "CUDA platform error: Error loading CUDA module: CUDA_ERROR_INVALID_PTX (218)" when I run python -m openmm.testInstallation on a GPU system. At first I did the install on the head node that has no GPUs and received the error, so then decided to do the install on node01 that has 4 - RTX 2080s, but am getting the same error. The cluster is running Bright Computing and I believe I have loaded the necessary modules.

From reading other posts on this issue, here are some of the details you may need. Any help would be greatly appreciated.
-I installed SEEKR via conda create --name SEEKR python=3.6.7
-I then activated SEEKR, then ran conda install -c conda-forge seekr2_openmm_plugin
-I was then able to run python -m openmm.testInstallation
Results:
(SEEKR) [root@node01 ~]# python -m openmm.testInstallation

OpenMM Version: 7.6
Git Revision: ad113a0cb37991a2de67a08026cf3b91616bafbe

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Error computing forces with CUDA platform
4 OpenCL - Successfully computed forces

CUDA platform error: Error loading CUDA module: CUDA_ERROR_INVALID_PTX (218)

Median difference in forces between platforms:

Reference vs. CPU: 6.29429e-06
Reference vs. OpenCL: 6.76294e-06
CPU vs. OpenCL: 8.071e-07

All differences are within tolerance.

Here are my GPU specs:
(SEEKR) [root@node01 ~]# nvidia-smi 
Fri Feb 18 19:49:10 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.01    Driver Version: 418.87.01    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 208...  On   | 00000000:3B:00.0 Off |                  N/A |
| 27%   28C    P8    21W / 250W |      0MiB /  7952MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce RTX 208...  On   | 00000000:5E:00.0 Off |                  N/A |
| 27%   32C    P8    13W / 250W |      0MiB /  7952MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce RTX 208...  On   | 00000000:AF:00.0 Off |                  N/A |
| 27%   33C    P8    27W / 250W |      0MiB /  7952MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce RTX 208...  On   | 00000000:D8:00.0 Off |                  N/A |
| 27%   31C    P8    20W / 250W |      0MiB /  7952MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

Any ideas on how to get this running with the GPUs??? It seems a lot of the issues are from the NVIDIA drivers not being updated. Do you think I should try updating the drivers next? The servers are running CentOS 7.7.1908.

(SEEKR) [root@node01 ~]# nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243

Thanks,
Joe Koral

**Comment by jchodera** [comment]
@peastman: Should we add this to the [FAQ](https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions)?

**Comment by jmkoral** [comment]
> That error means your CUDA driver is older than the CUDA toolkit. Here's your driver version:
> 
> > | NVIDIA-SMI 418.87.01 Driver Version: 418.87.01 CUDA Version: 10.1 |
> 
> Conda installs its own CUDA toolkit, and the oldest one it supports is 10.2. That's why you're getting the error. A workaround is to tell OpenMM to use the compiler from your 10.1 toolkit:
> 
> ```
> export OPENMM_CUDA_COMPILER=$(which nvcc)
> ```
> 
> Of course, if you can get the cluster administrators to upgrade to a newer driver, that would be better.

**Comment by jchodera** [comment]
CUDA 11.x are both newer than your driver version, which is the same problem as before. Can you use a CUDA version _equal to or older than_ your driver? You reported:

> | NVIDIA-SMI 418.87.01 Driver Version: 418.87.01 CUDA Version: 10.1 |

meaning you need to use CUDA 10.1 or earlier.

**Comment by jmkoral** [comment]
@jchodera: OK, I think it might be best to just update the NVIDIA drivers. The ticket I logged initially was using CUDA 10.1. It might be easiest to just update the drivers at this point. Thanks for all the quick responses.

**Comment by jmkoral** [comment]
This time I loaded CUDA 11, but did what peastman suggested and pointed my OpenMM compiler to point to version 10.1. I did receive a different error this time.

CUDA platform error: Error initializing FFT: 11

Thanks,
Joe

**Comment by peastman** [comment]
I'm not sure what that means.  According to the [API documentation](https://docs.nvidia.com/cuda/cufft/index.html#cufftresult), 11 is `CUFFT_INVALID_DEVICE`, which it describes as, "Execution of a plan was on different GPU than plan creation."  I don't think that's really what's happening.  I suspect it comes from linking to an incompatible version of cuFFT, for example an OpenMM package compiled against CUDA 10.2 trying to use a 10.1 toolkit at runtime.

**Comment by jmkoral** [comment]
Hey guys, I am still a little confused. You are suggesting I use CUDA 10.1 or earlier, which is what I am running. When I look at the release notes, I see CUDA 10.1 (10.1.105 general release, and updates) | >= 418.39, which is what I have.

NVIDIA-SMI 418.87.01    Driver Version: 418.87.01    CUDA Version: 10.1 
Cuda compilation tools, release 10.1, V10.1.243

I am the administrator of the cluster, but am really new to OpenMM and Bright Computing. So I am just trying to figure this all out. In order to update the NVIDIA drivers, it looks like I would have to update it on the Bright image and I don't think they want to do that right now.

**Comment by peastman** [comment]
I'm not familiar with Bright Computing.  In terms of CUDA compatibility, there are a few pieces involved.

1. The driver.  This is low level system software.  It's the piece that directly interacts with the hardware.  It's generally good to keep your driver up to date both to get bug fixes, and to support software that needs a recent driver.
2. The toolkit.  This is the code that programs interact with directly when using CUDA.  It's installed in user space, so anyone can download a toolkit and you can have multiple toolkits on your computer at once.  Each toolkit requires a minimum driver version.  Newer drivers can work with older toolkits, but not the other way around.
3. Programs that use CUDA (such as OpenMM).  A program is compiled against a particular version of the toolkit, and for the most part only works with that version.  They do have some limited binary compatibility between versions, but only within narrow ranges.

We distribute OpenMM through conda-forge, which creates packages compiled against a limited set of recent CUDA versions.  They add new versions as they come out, and drop old versions to keep the number of packages from getting too large.  Right now they support all CUDA versions from 10.2 on.

Here is what that means in practice.

- You can download OpenMM packages that were compiled against various CUDA versions, but nothing older than 10.2.
- A package compiled against 10.2 will not work with any toolkit older than 10.2.
- You can install a CUDA 10.2 toolkit, but it requires a minimum driver version.  Any driver older than that won't work.

**Comment by jmkoral** [comment]
@peastman: You have been great at replying so quickly. Just as a test, I decided to do the same install on a single GPU server running Ubuntu 18.04.6 LTS and am getting the same error. This server has Driver Version 440.64.00 with CUDA Version 10.2.

A little more background that I did not mention before. The user needs SEEKR as well, so I used the instructions provided on this page. https://seekr2.readthedocs.io/en/latest/installation.html#install-conda. All installs were done through Conda and I did not have to compile any software.

Should there be a different way I install the software so I can keep all the versions matched up?

**Comment by jmkoral** [comment]
I'm getting closer. I got it working on the single server with GPUs. Now I just have to figure it out on the cluster.

OpenMM Version: 7.7
Git Revision: 130124a3f9277b054ec40927360a6ad20c8f5fa6

There are 4 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces
3 CUDA - Successfully computed forces
4 OpenCL - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.30314e-06
Reference vs. CUDA: 6.71684e-06
CPU vs. CUDA: 7.47034e-07
Reference vs. OpenCL: 6.76294e-06
CPU vs. OpenCL: 8.05134e-07
CUDA vs. OpenCL: 2.65649e-07

All differences are within tolerance.

**Comment by jmkoral** [comment]
@peastman @jchodera I have been so close to getting this working. I finally understand what you have been referring to with the versions. With the Bright Computing software we are using for clustering, I am only able to load the latest CUDA that matches the version of the NVIDIA driver installed. So even if I load the module for CUDA11.0, the driver is still at 418, so can only load CUDA10.1. I am able to create different software images with Bright Computing so am testing it out on one node. After setting everything up, the node seems to take forever on Platform 3 when running python -m openmm.testInstallation. When I ssh into it, python is taking up 100% of the CPU. 

I think it is just a matter of matching up the versions like you said. I'll let you know what I figure out one I figure it out.

**Comment by jmkoral** [comment]
OK, sorry, it looks like I'm having the same issue on the cluster. Like above, the CPU on the node goes to 100% and just hangs. The other single GPU server I tested it on returned the results on the GPU within a couple seconds.

Any idea? And again, thanks for your time.

![Screenshot 2022-02-23_23-39-30-750](https://user-images.githubusercontent.com/58707121/155465153-50e66366-24dd-42a3-9866-ef4440f0681f.png)

**Comment by peastman** [comment]
Could you open a Python interpreter and execute these commands?  I'd like to see which one it's stuck on, CUDA or OpenCL.

```python
from openmm import *
print([Platform.getPlatform(i).getName() for i in range(Platform.getNumPlatforms())])
```

**Comment by jmkoral** [comment]
It does not look like it gets stuck at all when I run that. Here are the results and they were instant:
(SEEKR) [jkoral@node04 ~]$ python
Python 3.8.12 (default, Oct 12 2021, 13:49:34) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from openmm import *
>>> print([Platform.getPlatform(i).getName() for i in range(Platform.getNumPlatforms())])
['Reference', 'CPU', 'CUDA', 'OpenCL']
>>>

**Comment by peastman** [comment]
I'm guessing this is still a version incompatibility, leading to a miscommunication between libraries.  Anyway, I can't think of anything else that would do it.  We can see that CUDA is the platform it's stuck on.  Your output from `top` shows it's still using a full CPU core at nearly 100%, and it has been going for almost two minutes.  Compiling kernels is the only thing that should take significant CPU time, and that should be measured in seconds, not minutes.

**Comment by jchodera** [comment]
Is it possible the wrong version of a library is accidentally in the
LD_LIBRARY_PATH still?

**Comment by jmkoral** [comment]
I wanted to give you both an update on my situation. As it turns out, GPU0 in the cluster node is having issues. When running nvidia-smi, I noticed the GPU0 was running at 100% even when no processes were running. So I decided to try another node on the cluster and all the tests went 100% fine.

I wanted to thank you both for your responses. It has helped me learn more about OpenMM and a lot more about the cluster.
